File: LibraryAPI/daos/book_dao.py
from abc import ABC, abstractmethod
from typing import List

from entities.book import Book


class BookDAO(ABC):

    # CRUD
    # CREATE
    # create method SAVES a new book to a database or some other location
    @abstractmethod
    def create_book(self, book: Book) -> Book:
        pass

    # READ
    @abstractmethod
    def get_book_by_id(self, book_id: int) -> Book:
        pass

    @abstractmethod # List[Book] is the type annotation for saying a list of books
    def get_all_books(self) -> List[Book]:
        pass

    #UPDATE
    @abstractmethod
    def update_book(self, book: Book) -> Book:
        pass

    #DELETE
    @abstractmethod
    def delete_book(self, book_id: int) -> bool:
        pass

File: CommunityLibrariAPI/src/main/java/dev/ranieri/daos/BookDAO.java
package dev.ranieri.daos;

import dev.ranieri.entities.Book;
import dev.ranieri.exceptions.ResourceNotFound;

import java.util.List;

// Proper Interface for our DAO
public interface BookDAO {

    //CREATE
    Book createBook(Book book);

    //READ
    Book getBookById(int id);
    List<Book> getAllBooks(); // You should ALWAYS be using the interfaces

    //UPDATE
    Book updateBook(Book book);

    //DELETE
    boolean deleteBookById(int id);

}
